feat: 프로필 상태메시지 등록 및 수정 기능 구현

hgene0929 opened the pull request:
## Summary

* 프로필 등록시 이미지도 함께 등록돼야하지 않을까해서 ProfileImage 엔티티를 추가하고, 저장하는 로직도 함께 추가하였습니다.
  그 과정에서 기존 Image 클래스에서 uplaod() 의 void 반환값을 경로를 반환하도록 수정하였습니다.
* 아직 지우지는 않았지만 Profile 엔티티에 폰번호를 저장하지 않는다면, 관련 필드를 지워도 될까요??
* 테스트 코드는 내일 회의 전까지 작성해서 등록하도록 하겠습니다!!
  관련해서 serviceImpl 에 관한 테스트 코드에서 nullPointerException이 발생해서 확인하고 있는 중이기 때문입니다!!

*읽어주셔서 감사합니다.*

hgene2452 committed:
feat: 프로필 상태메시지 등록 및 수정 기능 구현

hgene2452 committed:
chore: 프로필 등록 기능

ndy2 commented on src/main/java/com/avalon/avalonchat/AvalonChatBackendApplication.java:
```diff
@@ -5,7 +5,7 @@
 
 @EnableJpaAuditing
-@SpringBootApplication
+@SpringBootApplication(scanBasePackages = {"com.avalon.avalonchat.domain.profile.domain"})
```
scanBasePackages 는 왜 추가되었나요?

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/controller/ProfileController.java:
```diff
@@ -1,30 +1,38 @@
 @RequiredArgsConstructor
 @Slf4j
-@RequestMapping("/profile")
+@RequestMapping("/v1/profiles")
```
#32 에 ServletContextPath 를 추가해서 v1 prefix 를 추가하지 않아도 될것 같습니다.
`/profiles`  면 충분할거 같네요

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/controller/ProfileController.java:
```diff
@@ -1,30 +1,38 @@
 
-	//TODO 프로필 닉네임 업데이트
+		//TODO jwt 토큰에서 user_id 꺼내기
+		long id = 1L;
```
이부분은 Spring Security 의 @AuthenticationPrincipal 을 활용하여 공통된 로직을 간소화 할 수 있습니다.

hgene2452 committed:
chore: 프로필 등록 응답 DTO 수정

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/domain/Profile.java:
```diff
@@ -1,34 +1,24 @@
 	private Long id;
 
-	@OneToOne(fetch = FetchType.LAZY)
+	@OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
```
cascade 가 추가된 이유는 무엇인가요?

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/dto/ProfileResponse.java:
```diff
@@ -0,0 +1,48 @@
+		String nickname,
+		String bio,
+		ProfileImage image,
+		ProfileImage backgroundImage) {
```
ProfileImage 클래스가 표현하고자 하는 의미는 프로필 이미지가맞을까요?
그렇다면 ProfileImage 클래스가 backGroundImage 를 파라미터로 받으면 상당히 헷갈릴거 같아요.

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/dto/ProfileRequest.java:
```diff
@@ -0,0 +1,46 @@
+	private String nickname;
+	private String bio;
+	private ImageRequest image;
+	private ImageRequest backgroundImage;
```
Dto 의 이름이 조금더 명확하게 지으면 좋을거 같아요.
이 Dto 는 프로필 등록을 위한 DTO 인데
생일을 받지 않는 프로필 업데이트 DTO 가 생긴다면 구분이 힘들거 같아요.

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/dto/image/ImageRequest.java:
```diff
@@ -0,0 +1,50 @@
+			byteArrayInputStream = new ByteArrayInputStream(byteArray);
+		} catch (IOException e) {
+			throw new ImageUploadException("파일시스템 이미지 업로드 중 예외 발생", e);
+		}
```
잘못된 예외 메시지가 추가되어 있어요.

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/controller/ProfileController.java:
```diff
@@ -1,30 +1,38 @@
+	public ResponseEntity<ProfileResponse> addProfile(
+		HttpServletRequest httpRequest,
+		@RequestBody ProfileRequest request
+	) {
```
아마 요 메서드 시그니처로는 멀티파트 폼 데이터 (이미지)를 잘 못받을 수도 있을거 같아요.
테스트가 시급합니당

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/dto/ProfileResponse.java:
```diff
@@ -0,0 +1,48 @@
+
+@Getter
+public class ProfileResponse {
+	private LocalDate birthDate;
```
응답의 경우도 마찬가지로 조금더 다양한 응답케이스를 고려해보면 ProfileResponse 는 조금은 포괄적인 이름같네요.

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/service/ProfileServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+import com.avalon.avalonchat.domain.user.domain.User;
+import com.avalon.avalonchat.domain.user.repository.UserRepository;
+import com.avalon.avalonchat.global.error.exception.AvalonChatRuntimeException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
```
서비스 계층에 `org.springframework.http` 에 대한 의존성이 추가되는 것은 좋지 않은 방식 같아요.

이동욱 님의 [이 글](https://jojoldu.tistory.com/603)을 한번 참고해보시면 좋을거 같아요.

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/service/ProfileServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+		return new ResponseEntity<>(profileResponse, HttpStatus.CREATED);
+	}
+
+	private HashMap<String, ProfileImage> uploadImage(ProfileRequest request, Profile profile) {
+		HashMap<String, ProfileImage> images = new HashMap<>();
+
+		ProfileImage profileImage = request.getImage().toEntity(profile, imageUploader);
+		images.put("image", profileImageRespository.save(profileImage));
+		ProfileImage profileBackgroundImage = request.getBackgroundImage().toEntity(profile, imageUploader);
+		images.put("backgroundImage", profileImageRespository.save(profileBackgroundImage));
+
+		return images;
+	}
```
uploadImage 라는 메서드네임에서 기대 되는 이미지 업로드 호출 (`Image.uploadBy(imageUploader)`) 이 이루어지지 않은것 같아요.

또한 해시맵 자료구조를 추가해 메서드를 뽑아내셧는데 이 부분이 조금은 읽기 어려워 보이네요.

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/dto/image/ImageRequest.java:
```diff
@@ -0,0 +1,50 @@
+		return file.getOriginalFilename();
+	}
+
+	public ProfileImage toEntity(Profile profile, ImageUploader imageUploader) {
+		String filePath = toImage().uploadBy(imageUploader);
+		return new ProfileImage(profile, filePath);
+	}
```
이 메서드의 이름과 역할이 조금 헷갈리네요.
ImageRequest - 이미지 요청에 대한 toEntity 호출로 ProfileImage 가 생성된다고 개인적으로 쉽게 예상이 잘 가지 않습니다.
또한 엔티티를 생성하는 과정에 ImageUploader 를 전달받아 이미지 업로드를 호출하는 것이 조금은 부적절해보이네요. 핵심 비즈니스 로직의 호출이 DTO 에 숨어있는 느낌입니당

참고) 흔히 이야기 하는 DTO/VO/Entity 관련 테이블입니다.
![image](https://user-images.githubusercontent.com/67302707/224657012-af0eb5c3-e2bf-462c-a644-c11f12a10ebc.png)

ndy2 reviewed (commented):
고생하셨습니다! 리뷰 남깁니다.

hgene2452 committed:
test: 프로필 등록 엔티티, 레포지토리, 서비스 테스트

hgene2452 committed:
fix: 프로필 등록 컨트롤러 수정

- prefix 수정: 공통 v1 제거
- userId 추출로직 수정: header에서 토큰 가져와서 userId 꺼내는 방식 -> 시큐리티 인증객체를 어노테이션 가지고 주입받아서 userId 바로 꺼내는 방식으로 변경할 예정(아직 소스 코드 풀 받기 전이라 TODO에 내용 변경해두었습니다)
- 요청 바디 바인딩 어노테이션 수정: multipart/form-data 바인딩 어노테이션 @ModelAttribute로 변경

** 추가로 컨트롤러 테스트 코드도 바로 짜서 커밋하도록 하겠습니다.

hgene2452 committed:
fix: 프로필 도메인/서비스 수정

- User(PK) 영속성관리 수정: cacade 사용 제거
- dto 이름 수정: dto 이름 명확하고 구체적으로 수정
- ByteArraInputStream 예외 수정: 글로벌 예외처리로
- Request dto의 toEntity() 메소드에 포함된 핵심 비즈니스 로직 추출해서 서비스 계층으로 분리
- 서비스 계층에서 http 의존성 제거

** ProfileImage가 표현하던 건 원래는 이미지(프로필,배경)의 엔티티가 맞습니다!! 그런데 혹시 해당 엔티티가 없어도 된다면,, 이미지 엔티티(경로)는 어떻게 저장되는지 알 수 있을까요?? 질문드려 죄송합니다!!
우선 해당 부분 관련 여쭙기 위해 구현에서 제외해놓았습니다..!

hgene2452 committed:
feat: ImageUploader 구현체 생성

 - ImageUploader 구현체를 생성하여 이미지 업로드 로직은 모두 infra.image.upload로 이동.
 - GCS 버킷 yml 통해 설정 후, 생성된 빈을 구현체에서 주입받아 내부 create() 메서드를 통해 업로드 기능 구현.
 - 이미지명은 uuid 방식 활용.

 ** 발급받은 키와 설정파일(yml)은 포함 안되어있습니다.
    어제 말씀 나눴던 방법 중 어떤 방법으로 외부에 공개되지 않도록 해야 할 지 몰라서 입니다..!!
 ** 곧바로 테스트 코드 작성해서 올리고, 추가적으로 프로필 등록도 마무리해서 커밋하겠습니다.

hgene2452 committed:
fix: Profile 등록 로직 수정

 - 메인애플리케이션: basescan 제거
 - BackgroundImage, ProfileImage: 배경이미지 프로필이미지 엔티티,레포지토리 생성
 - request dto: 프로필 dto 내부 필드에 바로 Image 타입 받을 수 있도록 수정
 - 서비스: 이미지 저장된 경로 받아와 객체 생성 후, response dto 생성하도록 완성

 ** 이미지 엔티티들 생성을 서비스단이 아니라, dto에서 해야 더 좋은 코드일 것 같긴한데, 이미지 업로드를 하는 서비스단에서 같은 프로필 서비스단으로 dto를 생성해서 보낼 필요가 있나하는 고민도 생겼었습니다.
 이에 현재는 서비스에서 new 로 바로 객체생성을 하게 되었습니다..!
 ** Map 같은 경우에는 우선 반환 타입에는 HashMap에서 Map 타입으로 수정하였는데, 이때 response dto를 생성하고 setter로 이미지 정보를 넣는 것보다는 Map 사용하는 것이 낫지 않을까하는 고민을 했었습니다..!

 ** 마찬가지로 테스트 코드 커밋하겠습니다. 또한 체크 통과 못하는 부분 및 혹시나 코멘트 있으시면 함께 반영해서 수정한 코드 주말 이전까지 커밋하도록 하겠습니다!!
 :fire: 감사합니다ㅎㅎ 좋은 주말 되세요!!

ndy2 commented on src/main/java/com/avalon/avalonchat/infra/image/upload/ImageUploader.java:
```diff
@@ -1,4 +1,6 @@
-package com.avalon.avalonchat.domain.profile.domain;
+package com.avalon.avalonchat.infra.image.upload;
```
이미지 업로더 인터페이스의 위치는 Domain 이 적절하다고 생각합니당.

그 이유는 이미지 업로더를 일종의 DDD 에서 이야기 하는 도메인 서비스 라고 생각하기 때문입니다.
저도 잘 모르고 어렵지만 [제가 전에 학습하고 정리한 요글](https://www.notion.so/ndy-dev/7-bb70530e43ff4243b98e838fbab1b0ca)을 훔쳐와서 인용해보면

> 도메인 서비스가 특정 기술에 의존하거나 외부 시스템의 API를 호출한다면 도메인 영역의 도메인 서비스는 인터> 페이스로 추상화 하고 Infrastructure 영역에서 구현해야 한다.
> 이를 통해 도메인 영역이 특정 구현에 종속되는 것을 방지할 수 있고 테스트 주도적인 개발이 가능하다.
라는 말이 있습니다. 요건 아마 DDD 시작하기라는 책의 최범균 저자님이 하신 말 같아요.

관련해서 저희가 말씀을 못 드렸던거 같은데 기본적으로 DDD 라는 방법론을 준수하는 방향으로 패키지 및 아키텍쳐 구조를 잡기로했습니다. 그래서 요 관련 내용도 조금 찾아보시면 도움이 될것 같아요. 

책으로는 위에서 이야기한 최범균 님의 `DDD 시작하기` 추천하고 글로는 M Yauri Attamimi 라는 분의 포스팅 시리즈 [@ Medium](https://yauritux.medium.com/) 추천드립니다.

포스팅 시리즈 중 DDD Part 1 ~ 3 을 이해하시면 ImageUploader를 왜 도메인에 넣었는지 의도가 조금은 와닿을 수 있을거에요.

ndy2 commented on src/main/java/com/avalon/avalonchat/infra/image/upload/ImageUploaderImpl.java:
```diff
@@ -0,0 +1,37 @@
+	private static final String URL_PREFIX = "https://storage.googleapis.com";
+	private final Storage storage;
+	@Value("${spring.cloud.gcp.storage.bucket}")
+	private String BUCKET_NAME;
```
Spring 의 `@ConfigurationProperties` 에 대해서 학습해보시고 
`com.google.cloud.spring.autoconfigure.storage.GcpStorageProperties` 요 클래스를 한번 찾아보시면 도움이 될것 같습니다.

ndy2 commented on src/main/java/com/avalon/avalonchat/infra/image/upload/ImageUploaderImpl.java:
```diff
@@ -0,0 +1,37 @@
+	private String fileWrite(Image image, String fileName) {
+		storage.create(
+			BlobInfo.newBuilder(BUCKET_NAME, fileName)
+				.setContentType(image.getExtension())
```
확장자(e.g. `jpg`)와 contentType(e.g. `images/jpeg`) 은 다릅니당.
저도 이런걸 고민해본적은 없는데 [여기](https://www.baeldung.com/java-file-mime-type) 에서 다양한 전환 방법들을 소개하네용.

ndy2 commented on src/main/java/com/avalon/avalonchat/infra/image/upload/ImageUploaderImpl.java:
```diff
@@ -0,0 +1,37 @@
+
+	private String fileWrite(Image image, String fileName) {
+		storage.create(
+			BlobInfo.newBuilder(BUCKET_NAME, fileName)
```
확인해보니 create 는 deprecated 이던데 createFrom 을 사용하면 문제가 있나요?

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/domain/image/Image.java:
```diff
@@ -1,5 +1,6 @@
-package com.avalon.avalonchat.domain.profile.domain;
+package com.avalon.avalonchat.domain.profile.domain.image;
 
+import com.avalon.avalonchat.infra.image.upload.ImageUploader;
```
여기 에서 ImageUploader 를 infra 로 보냇기 때문에 생기는 좋지않은 의존성 (domain -> infra) 이 추가되었네용.

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/domain/image/Image.java:
```diff
@@ -1,5 +1,6 @@
-package com.avalon.avalonchat.domain.profile.domain;
+package com.avalon.avalonchat.domain.profile.domain.image;
```
profile 과 image 가 성격상 어울리지 않는다고 판단하시고 별도의 패키지를 통해 관리하는 선택을 하셨군요.
뭔가 이상함을 느끼고 분리를 시도하는것은 좋다고 생각합니다. 하지만 요 방식 보다 개인적으로 DDD 도메인 영역에서의 역할 관점으로 나누는 것을 더 선호합니다. (보통 이런식으로 소개되기도 하고요). 

예를 들어
profile.domain.entity.Profile
profile.domain.entity.BackgroundImage
profile.domain.entity.ProfileImage
profile.domain.vo.Email (<- global 의 model 로 곧 보내야 할거 같긴 함)
profile.domain.serverice.ImageUploader
profile.domain.repository.ProfileRepository (<- 사실 DDD 에서 리포지토리도 도메인 영역으로 분류합니다. 근데 안 지키고 있었네용)
profile.domain.repository.BackGrounImageRepository (<-예네는 (여기) 있으면 안되긴합니당. 관련해서 이야기할 내용이 많긴 하네용)
profile.domain.repository.ProfileImageRepository

요렇게요.

사실 이렇게 도메인 레벨에서 image 라는 별도의 개념을 이야기하는 패키지까지 등장했다는 것은 `avalonchat.domain.image` 혹은 `avalonchat.domain.profileiamge`, `avalonchat.domain.backgroundimage` 라는 패키지를 추가하는 것을 고민해보아야 하는 신호같기도 하네용.

그리고 한 패키지에 클래스가 어느정도 등장해야 저런 구분을 주는가에도 개인차가 있는거 같습니다.

hgene2452 committed:
fix: Profile 등록 로직 수정

- ImageUploader 위치 이동: 외부기술(S3)에 의존하는 서비스의 인터페이스 위치를 DDD 전략에 따라 이동
- yml 프로퍼티 주입: .yml 파일에 위치한 프로퍼티를 Java 타입으로 관리하도록 프로퍼티 바인딩 클래스 생성
- 이미지 업로드 메소드 변경: deprecated 된 메소드 create() -> createFrom()
- 이미지 업로드 확장자 변경: Files.probeContentType() 활용
- 클래스 위치 이동: 기능별로 구성(이미지는 프로필에만 사용되기 때문에 프로필 이하로 이동)

ndy2 commented on src/main/java/com/avalon/avalonchat/global/configuration/GcpStorageProperties.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@Configuration
+@ConfigurationProperties(prefix = "spring.cloud.gcp.storage")
+public class GcpStorageProperties {
+	private final String bucket;
+	private final String url;
+}
```
제가 의미한 `GcpStorageProperties` 는`com.google.cloud.spring.autoconfigure.storage.GcpStorageProperties` 입니당. - [링크](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/storage/GcpStorageProperties.java)

Spring 자동 구성을 통해 빈으로 등록이 될태니 이것을 그대로 활용하면 어떨까요?

ndy2 commented on src/main/java/com/avalon/avalonchat/domain/profile/dto/ProfileAddRequest.java:
```diff
@@ -14,9 +14,9 @@
+	private final String nickname;
+	private final String bio;
 	private Image image;
 	private Image backgroundImage;
```
요청/응답 DTO 에서의 캡슐화, final 키워드의 사용은 Jackson 을 활용한 직렬화/역직렬화를 위해 희생되는 경우가 많습니다.

이를 해결하기 위해선 별도의 작업이 필요한데 이게 상당히 귀찮기 때문에 보통 팀적인 합의를 통해서 어느정도 감내하고 가는 편으로 알고 있어용. 때문에 이부분은 현재 크게 고민하지 않고 편하게 작성하시면 될것 같습니다.

ndy2 commented on build.gradle:
```diff
@@ -21,6 +21,7 @@ repositories {
 dependencies {
+    annotationProcessor "org.springframework.boot:spring-boot-configuration-processor"
     implementation group: 'org.springframework.cloud', name: 'spring-cloud-gcp-starter', version: '1.2.5.RELEASE'
     implementation group: 'org.springframework.cloud', name: 'spring-cloud-gcp-storage', version: '1.2.5.RELEASE'
```
요 의존성들도 다른 의존성 처럼 작성하면 더 나을거 같습니다.
```gradle
  implementation 'org.springframework.cloud:spring-cloud-gcp-starter:1.2.5.RELEASE'
  implementation 'org.springframework.cloud:spring-cloud-gcp-storage:1.2.5.RELEASE'
```

ndy2 commented on src/main/java/com/avalon/avalonchat/infra/image/upload/ImageUploaderImpl.java:
```diff
@@ -1,22 +1,25 @@
 	private final Storage storage;
-	@Value("${spring.cloud.gcp.storage.bucket}")
-	private String BUCKET_NAME;
+	private final GcpStorageProperties gcpStorageProperties;
```
매번 그런 것은 아니지만 요 경우에는
property 를 생성자에서 주입받아 필드에는 사용하는 값인 url 과 bucket 만을 보관해도 좋을거 같습니다.

ndy2 commented on src/main/java/com/avalon/avalonchat/infra/image/upload/ImageUploaderImpl.java:
```diff
@@ -1,22 +1,25 @@
 
 @Service
 @RequiredArgsConstructor
 public class ImageUploaderImpl implements ImageUploader {
```
클래스의 이름에 구체적인 사용 기술의 이름을 포함하면 어떨까요?

hgene2452 committed:
fix: Profile 테스트 코드 수정

hgene2452 committed:
fix: Profile 등록 코드 수정

- build.gradle: 의존성 주입 방식 통일
- ImageUploader 구현체:
  - 사용하는 구체적인 기술 이름 포함시킨 클래스명으로 변경
  - DI 방식 변경
- ProfileAddRequest: 내부 필드 final 키워드 제거

hgene2452 committed:
fix: Profile 등록 코드 수정

- ImageUploader 구현체: 사용하는 구체적인 기술 이름 포함시킨 클래스명으로 변경

ndy2 reviewed (approved):
고생하셨습니다! 💥
